src/hotspot/share/cds/cdsHeapVerifier.cpp

```diff
@@ -248,54 +248,63 @@ inline bool CDSHeapVerifier::do_entry(oop& orig_obj, HeapShared::CachedOopInfo&
 class CDSHeapVerifier::TraceFields : public FieldClosure {
   oop _orig_obj;
   oop _orig_field;
-  LogStream* _ls;
+  outputStream* _st;
 
 public:
-  TraceFields(oop orig_obj, oop orig_field, LogStream* ls)
-    : _orig_obj(orig_obj), _orig_field(orig_field), _ls(ls) {}
+  TraceFields(oop orig_obj, oop orig_field, outputStream* st)
+    : _orig_obj(orig_obj), _orig_field(orig_field), _st(st) {}
 
   void do_field(fieldDescriptor* fd) {
     if (fd->field_type() == T_OBJECT || fd->field_type() == T_ARRAY) {
       oop obj_field = _orig_obj->obj_field(fd->offset());
       if (obj_field == _orig_field) {
-        _ls->print("::%s (offset = %d)", fd->name()->as_C_string(), fd->offset());
+        _st->print("::%s (offset = %d)", fd->name()->as_C_string(), fd->offset());
       }
     }
   }
 };
 
+// Call this function (from gdb, etc) if you want to know why an object is archived.
+void CDSHeapVerifier::trace_to_root(outputStream* st, oop orig_obj) {
+  HeapShared::CachedOopInfo* info = HeapShared::archived_object_cache()->get(orig_obj);
+  if (info != NULL) {
+    trace_to_root(st, orig_obj, NULL, info);
+  } else {
+    st->print_cr("Not an archived object??");
+  }
+}
+
 // Hint: to exercise this function, uncomment out one of the ADD_EXCL lines above.
-int CDSHeapVerifier::trace_to_root(oop orig_obj, oop orig_field, HeapShared::CachedOopInfo* p) {
+int CDSHeapVerifier::trace_to_root(outputStream* st, oop orig_obj, oop orig_field, HeapShared::CachedOopInfo* info) {
```

Comment:
With this change, this function could also be exercised via a debugger. Should the comment be updated?

Reply:
Thanks for the review. Since the comment a few lines above here mentions gdb, this comment is out-dated, so I deleted it in the latest commit.